Mathlib/Topology/Connected/RealIntervals.lean

```diff
@@ -0,0 +1,560 @@
+  apply Subset.antisymm
+  · intro x xX
+    exact ⟨h_infX.1 xX,
+           x_lt_excluded_supX xX h_supX supX_X⟩
+  · rw [Ico_Ioo_with_endpoint inf_lt_sup, union_subset_iff]
+    exact ⟨connected_bdd_subset_contains_Ioo conn h_infX h_supX,
+           singleton_subset_iff.mpr infX_X⟩
+
+lemma characterize_Icc (infX_X : infX ∈ X) (supX_X : supX ∈ X) : X = Icc infX supX := by
+  apply Subset.antisymm
+  · exact bdd_subset_Icc h_infX h_supX
+  · exact (conn_has_Icc conn) infX infX_X supX supX_X
+
+lemma characterize_singleton (eq : infX = supX) : (X = {infX}) := by
+  rw [← eq] at h_supX
+  apply Subset.antisymm
+  · apply subset_trans (bdd_subset_Icc h_infX h_supX)
+    exact Set.Icc_subset {infX} rfl rfl
+  · rw [singleton_subset_iff]
+    rcases IsLUB.nonempty h_supX with ⟨x, xX⟩
+    have : infX = x := LE.le.antisymm (h_infX.1 xX) (h_supX.1 xX)
+    exact mem_of_eq_of_mem (id this) xX
+
+end inf_sup_section
+end conn_section
+end X_section
+
+/-- A tag for the tagged union of kinds of bounded intervals -/
+inductive BoundedIntervalKind | IooKind | IocKind | IcoKind | IccKind
+
+open BoundedIntervalKind
+
+/-- A tagged union of the kinds of nontrivial bounded intervals: `Ioo`, `Ioc`, `Ico`, and `Icc`. -/
+structure BoundedInterval :=
+  /-- the left endpoint (supremum) of the interval -/
+  left_endpoint : ℝ
+  /-- the right endpoint (infimum) of the interval -/
+  right_endpoint : ℝ
+  /-- a proof that the interval is nontrivial -/
+  left_lt_right : left_endpoint < right_endpoint
+  /-- tag indicating which endpoints are included -/
+  kind : BoundedIntervalKind
+
+/-- Convert the data of an interval to the set of real numbers it represents -/
+def BoundedInterval_as_set : BoundedInterval → Set ℝ
+  | ⟨a, b, _, IooKind⟩ => Ioo a b
+  | ⟨a, b, _, IocKind⟩ => Ioc a b
+  | ⟨a, b, _, IcoKind⟩ => Ico a b
+  | ⟨a, b, _, IccKind⟩ => Icc a b
+
+/-- True iff X is one of Ioo a b, Ioc a b, Ico a b, or Icc a b for some a < b-/
+def isBoundedInterval (X : Set ℝ) :=
+  ∃ I : BoundedInterval, X = BoundedInterval_as_set I
+
+lemma isBoundedInterval_Ioo {a b : ℝ} (lt : a < b) : isBoundedInterval (Ioo a b) := by
+  use ⟨a, b, lt, IooKind⟩; rw [BoundedInterval_as_set]
+lemma isBoundedInterval_Ioc {a b : ℝ} (lt : a < b) : isBoundedInterval (Ioc a b) := by
+  use ⟨a, b, lt, IocKind⟩; rw [BoundedInterval_as_set]
+lemma isBoundedInterval_Ico {a b : ℝ} (lt : a < b) : isBoundedInterval (Ico a b) := by
+  use ⟨a, b, lt, IcoKind⟩; simp only [BoundedInterval_as_set]
+lemma isBoundedInterval_Icc {a b : ℝ} (lt : a < b) : isBoundedInterval (Icc a b) := by
+  use ⟨a, b, lt, IccKind⟩; simp only [BoundedInterval_as_set]
+
+/-- The major lemma, assuming inf and sup exist. -/
+lemma classify_connected_reals_with_GLB_lt_LUB
+    {X : Set ℝ} (conn : IsConnected X) {infX supX : ℝ}
+    (h_infX : IsGLB X infX) (h_supX : IsLUB X supX) (lt : infX < supX) :
+    isBoundedInterval X := by
+  by_cases infX_X : infX ∈ X
+  · by_cases supX_X : supX ∈ X
+    · use ⟨infX, supX, lt, IccKind⟩
```

Comment:
Can this be written on one line, as `exact ⟨infX, supX, lt, IccKind, characterize_Icc conn h_infX h_supX infX_X supX_X⟩`? Same for the other three cases.